### ISSUE TITLE
New check: RedundantModifierCompactSource

### ISSUE THREAD

**Issue by surajgojanur** [checkstyle/checkstyle#20496]
Parent issue: #20226

Related issue: #20266

Related merged documentation PR: #20495

## Describe the problem in detail

JEP 512 compact source files allow fields and methods to be declared directly,
without writing an explicit enclosing class declaration.

A compact compilation unit implicitly declares a class that:

- is a final top-level class;
- has package access;
- belongs to the unnamed package;
- contains the fields and methods declared directly in the compact compilation
  unit.

Because of this special language context, some modifiers written on direct
compact-source members may be redundant or unnecessary.

The existing `RedundantModifierCheck` intentionally skips direct members of
compact source files. It continues to validate ordinary Java declarations,
including declarations inside ordinary classes nested in a compact source
file.

A separate check is proposed so compact-source-specific modifier rules can be
implemented without increasing the complexity of `RedundantModifierCheck`.

Proposed check name:

```text
RedundantModifierCompactSource
```

## Proposed configuration

The check has no properties.

```xml
<module name="Checker">
  <module name="TreeWalker">
    <module name="RedundantModifierCompactSource"/>
  </module>
</module>
```

All supported redundant modifiers are analyzed automatically.

The check is configured independently from `RedundantModifierCheck`. Users may
enable both checks when they want validation of:

- fields and methods declared directly in compact source files; and
- ordinary Java declarations, including declarations inside nested classes.

## Scope

`RedundantModifierCompactSource` validates only fields and methods declared
directly in a `COMPACT_COMPILATION_UNIT`.

It does not validate declarations inside:

- ordinary classes;
- nested classes;
- interfaces;
- annotation types;
- enums;
- records;
- anonymous classes.

Those declarations remain in the scope of `RedundantModifierCheck`.

### Redundant `final` modifier on a direct method

The class implicitly declared by a compact compilation unit is final.
Therefore, an instance method declared directly in that implicit class cannot
be overridden by a subclass.

```java
public final void myMethod() { } // violation on 'final'

void main() { }
```

Expected output:

```text
Redundant 'final' modifier on a direct member of a compact source file.
```

### Redundant `public` modifier on a direct method

The implicit class has package access, belongs to the unnamed package, and
cannot be used as an ordinary public API type from another compilation unit.

```java
public void publicMethod() { } // violation on 'public'

void main() { }
```

Expected output:

```text
Redundant 'public' modifier on a direct member of a compact source file.
```

### Redundant `protected` modifier on a direct method

The implicit class is final and cannot be subclassed.

```java
protected void protectedMethod() { } // violation on 'protected'

void main() { }
```

Expected output:

```text
Redundant 'protected' modifier on a direct member of a compact source file.
```

### Redundant `private` modifier on a direct method

A direct compact-source method cannot be exposed through a normally named
public class declaration.

```java
private void privateMethod() { } // violation on 'private'

void main() { }
```

Expected output:

```text
Redundant 'private' modifier on a direct member of a compact source file.
```

### Visibility modifiers on direct fields

Visibility modifiers on fields declared directly in a compact compilation unit
should also be analyzed.

```java
public int first;       // violation on 'public'
protected int second;   // violation on 'protected'
private int third;      // violation on 'private'

void main() { }
```

Expected output:

```text
Redundant 'public' modifier on a direct member of a compact source file.
Redundant 'protected' modifier on a direct member of a compact source file.
Redundant 'private' modifier on a direct member of a compact source file.
```

### Multiple redundant modifiers

Each redundant modifier should receive its own violation.

```java
public final void process() { }
        // violations on 'public' and 'final'

void main() { }
```

Expected output:

```text
Redundant 'public' modifier on a direct member of a compact source file.
Redundant 'final' modifier on a direct member of a compact source file.
```

## Nested ordinary classes are excluded

A named class declared inside a compact source file is an ordinary class
declaration. Its members are not direct fields or methods of the compact
compilation unit for the purpose of this check.

```java
final class Nested {
    public final void method() { }
}

void main() { }
```

`RedundantModifierCompactSource` must not report violations inside `Nested`.

`RedundantModifierCheck` remains responsible for validating the nested class
and its members.

## Proposed message

The new check should define its own message instead of reusing the message
constant from `RedundantModifierCheck`.

Possible message entry:

```properties
redundantModifierCompactSource=Redundant ''{0}'' modifier on a direct member of a compact source file.
```

A separate message allows the violation to include compact-source-specific
context and allows its wording to evolve independently from
`RedundantModifierCheck`.

## Rationale

`RedundantModifierCheck` already handles many unrelated Java declaration
contexts and modifier rules.

Direct members of compact source files have a special language context because
their enclosing class is implicit, final, package-access, and located in the
unnamed package.

A separate check provides:

- a clearly defined compact-source-only scope;
- independent configuration;
- no behavior change for existing `RedundantModifierCheck` users;
- compact-source-specific messages;
- no additional properties;
- the ability for users to enable either or both checks.

## Documentation relationship

The documentation for `RedundantModifierCompactSource` should explain that it
validates only fields and methods declared directly in compact source files. It
should reference `RedundantModifierCheck` for ordinary and nested declarations.

Once the new check is implemented, the documentation for
`RedundantModifierCheck` should reference `RedundantModifierCompactSource` for
direct compact-source members.

## Confirmed design decisions

- Only direct fields and methods of compact source files are in scope.
- Ordinary and nested classes remain under `RedundantModifierCheck`.
- The check has no properties.
- Supported modifiers are analyzed automatically.
- The check is configured independently.
- The check defines its own messages.
- Both checks are documented as complementary checks.

## Clarification for `static`

`static` was identified as one of the target modifiers.

However, under the Java language rules, `static` changes a method from an
instance method to a class method and changes a field from per-instance storage
to one class-level value.

Please confirm whether:

1. every `static` modifier on a direct compact-source field or method should
   produce a violation as unnecessary compact-source style; or
2. `static` should produce a violation only in a narrower context where it is
   semantically redundant.

After this is confirmed, explicit `static` examples and expected output can be
added to the design.

**Comment by romani** [comment]
> Should this check cover only top-level methods, or both top-level methods and fields?

only top level, nested classes shold be covered RedundantModifierCheck, we can reference this in documentation.

> Which modifiers should be checked first: final, visibility modifiers, static, or a configurable subset?

no properties in Check, all of them are target: static. final, public , protected, ... 

> Should nested ordinary classes inside compact source files stay out of scope and continue to be handled by existing checks?

yes by RedundantModifierCheck

> Should this check reuse RedundantModifierCheck messages or define compact-source-specific messages?

no, better to not reuse, live showed us that Checks likely want to extend message a bit more in certain context, common message make it impossible. Copy paste is cheaper than dependency.

> Should this check be enabled separately from RedundantModifierCheck?

yes, all Checks are fully independent , and clearly define scope of validations. User do mix and match to archive what user wants.

> Should RedundantModifierCheck and RedundantModifierCompactSource be documented as complementary checks?

yes, we can reference in description that scope is limited and other part is done by other Check.

**Comment by romani** [comment]
@surajgojanur , please update issue description to follow this new module pattern https://github.com/checkstyle/checkstyle/blob/master/docs/report-issue.md#how-to-request-new-checkmodule

**Comment by surajgojanur** [comment]
> [@surajgojanur](https://github.com/surajgojanur) , please update issue description to follow this new module pattern https://github.com/checkstyle/checkstyle/blob/master/docs/report-issue.md#how-to-request-new-checkmodule

@romani  I updated the issue description to follow the official new
Check/Module request pattern.

It now includes the proposed configuration, direct-member scope, expected
violations, nested-class exclusion, separate message, rationale, and
documentation relationship.

I also added a precise clarification for `static`, since it was listed as a
target modifier but changes instance/class semantics under the JLS.

**Comment by Carbon14-48** [comment]
hi @surajgojanur are you gonna work on this issue ?

**Comment by surajgojanur** [comment]
> hi [@surajgojanur](https://github.com/surajgojanur) are you gonna work on this issue ?

Yes, I plan to work on this issue.

@romani  has clarified the main scope and design. I am currently waiting
for confirmation about the exact behavior of the `static` modifier, and after
that I will start the implementation

**Comment by Carbon14-48** [comment]
> > hi [@surajgojanur](https://github.com/surajgojanur) are you gonna work on this issue ?
> 
> Yes, I plan to work on this issue.
> 
> @romani  has clarified the main scope and design. I am currently waiting
> for confirmation about the exact behavior of the `static` modifier, and after
> that I will start the implementation

Goood luckkk !!

**Comment by romani** [comment]
static should be also violated

**Comment by surajgojanur** [comment]
@romani , before I begin implementation, I would like to confirm the exact initial modifier matrix so I do not expand the scope incorrectly.

Proposed method modifiers:
- public
- protected
- private
- static
- final

Proposed field modifiers:
- public
- protected
- private
- static

Could you please confirm:

1. Should `final` fields remain excluded, since `final` changes field  reassignment semantics?

2. Should `strictfp` methods be handled by this new check, or should they remain outside its scope?

3. For `static`, should the message retain the proposed wording:

   `Redundant ''{0}'' modifier on a direct member of a compact source file.`

   or should it use wording such as `Unnecessary` or `Disallowed`, since
   `static` changes method/field semantics?

After this confirmation, I will implement the new check with only the approved modifier sets.

**Comment by romani** [comment]
> Should final fields remain excluded, since final changes field reassignment semantics?

keep final for fields, as final make it constant, this is good modifier.

>Should strictfp methods be handled by this new check, or should they remain outside its scope?

add it , it is not required even for methods in modern jdk.
`Java 17 and newer: The strictfp modifier is completely obsolete and redundant. Since [JEP 306](https://openjdk.org/jeps/306),`

> For static, should the message retain the proposed wording:

`Redundant ''{0}'' modifier on a direct member` no need to mention one more time that it is compact source.  
static is redundant.
